json/src/machine.nr

```diff
@@ -0,0 +1,500 @@
+///
+/// # Returns:
+/// An array of three Field values representing masks for:
+/// - [0]: read_write_value mask (whether to modify the stack)
+/// - [1]: parsing_string mask (whether to toggle string parsing)
+/// - [2]: parsing_primitive mask (whether to toggle primitive parsing)
+fn state_to_mask(
+    read_delimeter: bool,
+    read_primitive: bool,
+    parsing_string: bool,
+    parsing_primitive: bool,
+) -> [Field; 3] {
+    // Convert booleans to Field values explicitly to avoid issues with boolean operations
+    let parsing_string_field = parsing_string as Field;
+    let parsing_primitive_field = parsing_primitive as Field;
+    let read_delimeter_field = read_delimeter as Field;
+    let read_primitive_field = read_primitive as Field;
+
+    // Calculate masks using Field arithmetic
+    // Mask for read_write_value: only modify stack if not parsing a string
+    let mask_rw = 1 - parsing_string_field;
+
+    // Mask for parsing_string: toggle between 1 and 0 to flip the bit
+    let mask_ps = 1 - parsing_string_field;
+
+    // Complex logic for parsing_primitive mask:
+
+    //  Currently parsing a primitive and read a delimiter -> end primitive parsing
+    let parsing_number_read_delim = parsing_primitive_field * read_delimeter_field;
+
+    // Not parsing a primitive but read a primitive character -> start primitive parsing
+    let read_prim_not_parsing = (1 - parsing_primitive_field) * read_primitive_field;
+
+    // Only apply these cases if not parsing a string
+    let part1 = (1 - parsing_string_field) * (parsing_number_read_delim + read_prim_not_parsing);
+
+    //  Currently parsing a primitive, read a non-primitive, non-delimiter character
+    // -> continue parsing primitive (e.g., whitespace in a number)
+    let part2 = parsing_primitive_field * (1 - read_primitive_field) * (1 - read_delimeter_field);
+
+    [mask_rw, mask_ps, part1 + part2]
+}
+
+/// Gets the top element of the stack and its position.
+/// The stack represents the current JSON structure being parsed
+pub fn get_top_of_stack(stack: [[Field; 2]; MAX_STACK_HEIGHT]) -> ([Field; 2], Field) {
+    let mut value: [Field; 2] = [Field::from(0), Field::from(0)];
+    let mut selector: Field = Field::from(0);
+
+    // Count non-zero entries and keep track of values
+    for i in 0..MAX_STACK_HEIGHT {
+        // Check if stack[i] is not [0,0] (i.e., it's allocated/used)
+        let is_allocated = stack[i] != [Field::from(0), Field::from(0)];
+
+        // Increment selector if slot is allocated
+        if is_allocated {
+            selector += Field::from(1);
+        }
+
+        // Store this value if it's the highest allocated slot
+        // This effectively finds the top of the stack
+        if is_allocated {
+            value = stack[i];
+        }
+    }
+
+    (value, selector)
+}
+
+/// Rewrites the stack and tree hash based on the current byte and parsing state.
+pub fn rewrite_stack(input: RewriteStackInput) -> RewriteStackOutput {
```

Comment:
You can use like `rewrite_stack<T>` here.
https://noir-lang.org/docs/dev/noir/concepts/generics